crates/sui-storage/src/event_store.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! EventStore API supports high velocity event ingestion plus flexible event querying.
+//! Multiple use cases supported:
+//! - Explorer reads of different events
+//! - Filtering of events per Move package, type, or other fields
+//! - Persistent/reliable streaming, which needs to recover filtered events from a marker
+//!   or point in time
+//!   
+//! Events are also archived into checkpoints so this API should support that as well.
+//!
+
+use move_core_types::language_storage::ModuleId;
+use sui_types::base_types::TransactionDigest;
+use sui_types::event::EventEnvelope;
+
+use flexstr::SharedStr;
+use serde_json::Value;
+
+/// One event pulled out from the EventStore
+#[allow(unused)]
+struct StoredEvent {
+    /// UTC timestamp in milliseconds
+    timestamp: u64,
+    checkpoint_num: u64,
+    /// Only present for events pertaining to specific transactions
+    tx_digest: Option<TransactionDigest>,
+    /// The variant name from SuiEvent, eg MoveEvent, Publish, etc.
+    event_type: SharedStr,
+    /// Will be None for System events
+    move_module: Option<SharedStr>,
+    /// Individual event fields.  As much as possible these should be deconstructed and flattened,
+    /// ie `{'obj': {'fieldA': 'A', 'fieldB': 'B'}}` should really be broken down to
+    // `[('obj.fieldA', 'A'), ('obj.fieldB', 'B')]
+    fields: Vec<(SharedStr, Value)>, // Change this to something based on CBOR for binary values, or our own value types for efficiency
+}
+
+/// An EventStore supports event ingestion and flexible event querying
+/// One can think of events as logs.  They represent a log of what is happening to Sui.
+/// Thus, all different kinds of events fit on a timeline, and one should be able to query for
+/// different types of events that happen over that timeline.
+trait EventStore<EventIt>
+where
+    EventIt: Iterator<Item = StoredEvent>,
+{
+    /// Adds events to the EventStore.
+    /// Semantics: events are appended, no deduplication is done.
+    /// All events input must belong to the same checkpoint number.
+    fn add_events(
+        &self,
+        events: &[EventEnvelope],
+        checkpoint_num: u64,
+    ) -> Result<(), EventStoreError>;
+
+    /// Queries for events emitted by a given transaction, returned in order emitted
+    /// NOTE: Not all events come from transactions
+    fn events_for_transaction(&self, digest: TransactionDigest)
+        -> Result<EventIt, EventStoreError>;
```

Comment:
do we want to return Derived Events from this api? if so, maybe need 2 `EventIt` or add a flag etc

Reply:
I don't consider derived events as separate.  They should be returned by the same API.  All events are just events, you can always filter for derived events if you really want

Reply:
Let's use a specific example.  Let's say we add an event for a deleted object.  This is inserted automatically when we finish processing a transaction.  If a user move event is emitted, it would also be added.  They would all appear in the events of a `TransactionEffects` together.  The order within the transaction, I don't think it matters because they are all generated by the same transaction.  (This is one reason I want to generate as much as possible as close to when the effects are produced as possible)

Reply:
resolved offline